### PR TITLE
Add a switch to choose output version for audit events

### PR DESCRIPTION
**Pull request by soltysh**
This is followup to https://github.com/kubernetes/kubernetes/pull/51719 to start the discussion how we want to solve the problem of users picking which version is being served them.

We need to have an option for log and webhook, separately. Probably, for webhook backend with multiple destinations we'd like to send different version to each.

This approach adds two flags (only the second commit matters), one for log and another for webhook (unfortunately global one). I've looked into [kubeconfig types](https://github.com/kubernetes/kubernetes/blob/6a845c67f097cd76f9d90ab3c9c5b5603c70afe4/staging/src/k8s.io/client-go/tools/clientcmd/api/types.go#L30-L37) and although there are options to specify group and version they are meant for removal. @liggitt had some thoughts maybe he could share the ideas and we can pick it up here.

@ericchiang @CaoShuFeng @crassirostris @sttts opinions, thoughts are more than welcome

**Release note**:
```release-note
Add option to choose audit output version.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go [hunk 1]
```diff
@@ -66,6 +67,9 @@ type AuditLogOptions struct {
 	MaxBackups int
 	MaxSize    int
 	Format     string
+	// Preferred group version for the log output.
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: Same comment as in https://github.com/kubernetes/kubernetes/pull/51719#discussion_r136378611, word "preferred" sounds little bit confusing

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go [hunk 2]
```diff
@@ -136,6 +149,17 @@ func (o *AuditOptions) Validate() []error {
 	return allErrors
 }
 
+func validateGroupVersionString(groupVersion string) error {
+	gv, err := schema.ParseGroupVersion(groupVersion)
+	if err != nil {
+		return err
+	}
+	if gv != auditv1beta1.SchemeGroupVersion && gv != auditv1alpha1.SchemeGroupVersion {
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: Nit: move to the list of allowed versions?

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go (hunk 1 above)
Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: Same in webhook options and in flag descriptions

**Review by crassirostris — COMMENTED** [review]
Thanks a lot for the PR!

**Comment by ericchiang** [comment]
I'd be fine (ab)using the kubeconfig fields. Something like

```
# users refers to the API server's webhook configuration.
users:
  - name: name-of-api-server
    user:
      client-certificate: /path/to/cert.pem # cert for the webhook plugin to use
      client-key: /path/to/key.pem          # key matching the cert
    apiVersion: audit.k8s.io/v1alpha1
```

This would also translate well to all webhook and `k8s.io/apiserver/pkg/util/webhook`. E.g. something like

```go
// NewGenericWebhook ...
//
// groupVersions represent the set of supported serialization versions for the resource. If the kubeconfig
// specifies an explicit apiVersion that groupVersion is used. If not, the first version in the list is used. 
func NewGenericWebhook(registry *registered.APIRegistrationManager, codecFactory serializer.CodecFactory, kubeConfigFile string, groupVersions []schema.GroupVersion, initialBackoff time.Duration) (*GenericWebhook, error)
```

Right now we never supply more than one groupVersion for any of our existing webhooks.

**Comment by crassirostris** [comment]
@soltysh Will you have time until codefreeze to work on this?

**Comment by crassirostris** [comment]
I'll take this PR over then

**Pull request closed without merging**